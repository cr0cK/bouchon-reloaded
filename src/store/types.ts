export type State = Record<string, any>

export type ActionId = string

export type Action = () => {
  name: string
  id: ActionId
}

export type Reducer = () => State

export type Reducers = Map<ActionId, Reducer>
