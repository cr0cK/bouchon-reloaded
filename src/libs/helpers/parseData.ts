import * as fs from 'fs'
import { newLogger } from './logger'

/**
 * Parse JSON and verify integrify via a JSON schema.
 *
 * Usage:
 *
 * const { ... } = createCork<...>(
 *   parseData('users-data.json')
 * )
 */
export function parseData<TStore>(dataPathname: string): TStore {
  const logger = newLogger('parseJSONAsStore')

  try {
    // TODO Use AJV to verify JSON integrity

    return JSON.parse(fs.readFileSync(dataPathname, 'utf8')) as TStore
  } catch (err) {
    logger.error(`Can't parse ${dataPathname}, exiting.`, err)
    process.exit(1)
  }
}
