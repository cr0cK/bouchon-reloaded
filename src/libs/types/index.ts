import * as express from 'express'

/** Helpers */

export type Perhaps<T> = T | null | undefined
export type MaybeUndef<T> = T | undefined
export type Maybe<T> = T | null

/** State machine */

export type State = Record<string, any>

export interface Action {
  name: string
  parameters?: object
  bodyParameters?: object
  headerParameters?: object
  queryParameters?: object
}

export type ActionFn<A> = (action: A) => void

export type Reducer<TState extends State, TActionProps> = (
  state: TState,
  action: TActionProps
) => TState

export type Reducers<
  TState extends State,
  TActionEnum extends string,
  TActionsRecord extends Record<TActionEnum, object>
> = Map<keyof TActionsRecord, Reducer<TState, any>>

/** Selectors */

export type Selector<TState, TActionUnion, TSelectorReturn> = (
  state: TState,
  action: TActionUnion
) => TSelectorReturn

export type SelectorFn<TActionUnion, TSelectorReturn> = (
  action: TActionUnion
) => TSelectorReturn

/** Routing */

export type Route<TActionUnion, TSelectorReturn> = {
  method: 'GET' | 'POST' | 'PATCH' | 'PUT' | 'DELETE' | 'HEAD'
  pathname: string
  action: ActionFn<any>
  selector?: SelectorFn<TActionUnion, TSelectorReturn>
  handler?: (
    selectedData: TSelectorReturn,
    action: TActionUnion
  ) => (req: express.Request, res: express.Response) => void
}

export type AnyRoute = Route<any, any>

export type EndPoint = {
  pathname: string
  routes: Array<Route<any, any>>
}

/** Types helpers */

export interface IMockIdRelation {
  mockId: MockId
  parentMockId: MaybeUndef<MockId>
}

// useful to identify generated items
export type MockId = number
