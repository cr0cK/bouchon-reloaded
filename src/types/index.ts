/** State machine */

export type State = Record<string, any>

export interface ActionProps {
  name: string
  parameters?: object
  bodyParameters?: object
  headerParameters?: object
}

export type Reducer<TState extends State, TActionProps> = (
  state: TState,
  actionProps: TActionProps
) => TState

export type Reducers<
  TState extends State,
  TActionEnum extends string,
  TActionsRecord extends Record<TActionEnum, object>
> = Map<keyof TActionsRecord, Reducer<TState, any>>

/** Selectors */

export type SelectorFn<S, AP, R> = (state: S, actionProps: AP) => R

export type Selector<TState, AP, R> = (
  selectorFn: SelectorFn<TState, AP, R>
) => R

/** Routing */

export type Route<TState, AP, R> = {
  method: 'GET' | 'POST' | 'PATH' | 'PUT' | 'DELETE' | 'HEAD'
  pathname: string
  selector: Selector<TState, AP, R>
}
