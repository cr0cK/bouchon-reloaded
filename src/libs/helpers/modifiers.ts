import { merge } from 'lodash'
import { Maybe, MaybeUndef } from '../types'

type UpdateMapValuePredicateFn<K, V> = ([k, v]: [K, V]) => boolean

/**
 * Update a value of an array.
 */
export function updateArrayValue<T>(
  array: T[],
  searchPredicate: (value: T) => boolean,
  value: Partial<T>
): T[] {
  const index = array.findIndex(searchPredicate)

  if (index < 0) {
    return array
  }

  array.splice(index, 1, merge(array[index], value))

  return array
}

/**
 * Update a value of a map.
 */
export function updateMapValue<K extends string | number, V>(
  map: Map<K, V>,
  keyOrSearchPredicate: K | UpdateMapValuePredicateFn<K, V>,
  value: Partial<V>
): Map<K, V> {
  const entry = _getEntry(map, keyOrSearchPredicate)

  if (!entry) {
    return map
  }

  const [entryKey, entryValue] = entry

  map.set(entryKey, merge(entryValue, value))

  return map
}

/**
 * Delete a value from a map.
 */
export function deleteMapValue<K extends string | number, V>(
  map: Map<K, V>,
  keyOrSearchPredicate: K | UpdateMapValuePredicateFn<K, V>
): Map<K, V> {
  const entry = _getEntry(map, keyOrSearchPredicate)

  if (!entry) {
    return map
  }

  const [entryKey] = entry

  map.delete(entryKey)

  return map
}

function _getEntry<K extends string | number, V>(
  map: Map<K, V>,
  keyOrSearchPredicate: K | UpdateMapValuePredicateFn<K, V>
): MaybeUndef<[K, V]> {
  if (typeof keyOrSearchPredicate === 'function') {
    return Array.from(map.entries()).find(keyOrSearchPredicate)
  }

  const entry = map.get(keyOrSearchPredicate)

  if (!entry) {
    return
  }

  return [keyOrSearchPredicate, entry]
}
