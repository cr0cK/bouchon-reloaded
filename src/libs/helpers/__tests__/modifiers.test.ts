import { MockId } from '../../types'
import { deleteMapValue, updateArrayValue, updateMapValue } from '../modifiers'
import { stringifyData } from '../parseData'

interface IUser {
  name: string
  email: string
}

describe('modifiers', () => {
  describe('updateArrayValue', () => {
    let store: { users: Array<IUser & { id: number }> }

    beforeEach(() => {
      store = {
        users: [
          {
            id: 1,
            name: 'Bob',
            email: 'bob@aol.com'
          },
          {
            id: 2,
            name: 'Alice',
            email: 'alice@aol.com'
          }
        ]
      }
    })

    it('should update a value of the map', () => {
      const searchedEmail = 'bob@aol.com'

      updateArrayValue(store.users, value => value.email === searchedEmail, {
        name: 'New Bob'
      })

      const str = stringifyData(store)
      expect(str).toMatchSnapshot()
    })

    it('shouldnt update a value if the value is not found', () => {
      const searchedEmail = 'notfound@aol.com'

      updateArrayValue(store.users, value => value.email === searchedEmail, {
        name: 'New Bob'
      })

      const str = stringifyData(store)
      expect(str).toMatchSnapshot()
    })
  })

  describe('updateMapValue', () => {
    let store: { users: Map<MockId, IUser> }

    beforeEach(() => {
      store = {
        users: new Map([
          [
            1,
            {
              name: 'Bob',
              email: 'bob@aol.com'
            }
          ],
          [
            2,
            {
              name: 'Alice',
              email: 'alice@aol.com'
            }
          ]
        ])
      }
    })

    it('should update a value of the map with a predicate fn', () => {
      const searchedEmail = 'bob@aol.com'

      updateMapValue(
        store.users,
        ([, value]) => value.email === searchedEmail,
        {
          name: 'New Bob'
        }
      )

      const str = stringifyData(store)
      expect(str).toMatchSnapshot()
    })

    it('should update a value of the map with a defined key', () => {
      updateMapValue(store.users, 1, {
        name: 'New Bob'
      })

      const str = stringifyData(store)
      expect(str).toMatchSnapshot()
    })

    it('shouldnt update a value if the value is not found', () => {
      const searchedEmail = 'notfound@aol.com'

      updateMapValue(
        store.users,
        ([, value]) => value.email === searchedEmail,
        {
          name: 'New Bob'
        }
      )

      updateMapValue(store.users, -1, {
        name: 'New Bob'
      })

      const str = stringifyData(store)
      expect(str).toMatchSnapshot()
    })
  })

  describe('deleteMapValue', () => {
    let store: { users: Map<MockId, IUser> }

    beforeEach(() => {
      store = {
        users: new Map([
          [
            1,
            {
              name: 'Bob',
              email: 'bob@aol.com'
            }
          ],
          [
            2,
            {
              name: 'Alice',
              email: 'alice@aol.com'
            }
          ]
        ])
      }
    })

    it('should delete a value of the map with a defined key', () => {
      deleteMapValue(store.users, 1)

      const str = stringifyData(store)
      expect(str).toMatchSnapshot()
    })

    it('should delete a value of the map with a predicate fn', () => {
      const searchedEmail = 'bob@aol.com'

      deleteMapValue(store.users, ([, value]) => value.email === searchedEmail)

      const str = stringifyData(store)
      expect(str).toMatchSnapshot()
    })
  })
})
