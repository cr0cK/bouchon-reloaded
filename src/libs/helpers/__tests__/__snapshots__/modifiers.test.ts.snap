// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modifiers deleteMapValue should delete a value of the map with a defined key 1`] = `
"{
  "users": {
    "dataType": "Map",
    "value": [
      [
        2,
        {
          "name": "Alice",
          "email": "alice@aol.com",
          "roles": [
            "user"
          ]
        }
      ]
    ]
  }
}"
`;

exports[`modifiers deleteMapValue should delete a value of the map with a predicate fn 1`] = `
"{
  "users": {
    "dataType": "Map",
    "value": [
      [
        2,
        {
          "name": "Alice",
          "email": "alice@aol.com",
          "roles": [
            "user"
          ]
        }
      ]
    ]
  }
}"
`;

exports[`modifiers updateArrayValue should not update a value if the value is not found 1`] = `
"{
  "users": [
    {
      "id": 1,
      "name": "Bob",
      "email": "bob@aol.com",
      "roles": [
        "user"
      ]
    },
    {
      "id": 2,
      "name": "Alice",
      "email": "alice@aol.com",
      "roles": [
        "user"
      ]
    }
  ]
}"
`;

exports[`modifiers updateArrayValue should update a value of the map 1`] = `
"{
  "users": [
    {
      "id": 1,
      "name": "New Bob",
      "email": "bob@aol.com",
      "roles": [
        "user"
      ]
    },
    {
      "id": 2,
      "name": "Alice",
      "email": "alice@aol.com",
      "roles": [
        "user"
      ]
    }
  ]
}"
`;

exports[`modifiers updateMapValue with a defined key should not update a value if the value is not found 1`] = `
"{
  "users": {
    "dataType": "Map",
    "value": [
      [
        1,
        {
          "name": "Bob",
          "email": "bob@aol.com",
          "roles": [
            "user"
          ]
        }
      ],
      [
        2,
        {
          "name": "Alice",
          "email": "alice@aol.com",
          "roles": [
            "user"
          ]
        }
      ]
    ]
  }
}"
`;

exports[`modifiers updateMapValue with a defined key should update a value of the map 1`] = `
"{
  "users": {
    "dataType": "Map",
    "value": [
      [
        1,
        {
          "name": "New Bob",
          "email": "bob@aol.com",
          "roles": [
            "user"
          ]
        }
      ],
      [
        2,
        {
          "name": "Alice",
          "email": "alice@aol.com",
          "roles": [
            "user"
          ]
        }
      ]
    ]
  }
}"
`;

exports[`modifiers updateMapValue with a defined key should update list values by concatenation 1`] = `
"{
  "users": {
    "dataType": "Map",
    "value": [
      [
        1,
        {
          "name": "New Bob",
          "email": "bob@aol.com",
          "roles": [
            "user",
            "admin"
          ]
        }
      ],
      [
        2,
        {
          "name": "Alice",
          "email": "alice@aol.com",
          "roles": [
            "user"
          ]
        }
      ]
    ]
  }
}"
`;

exports[`modifiers updateMapValue with a defined key should update list values by overriding 1`] = `
"{
  "users": {
    "dataType": "Map",
    "value": [
      [
        1,
        {
          "name": "New Bob",
          "email": "bob@aol.com",
          "roles": [
            "admin"
          ]
        }
      ],
      [
        2,
        {
          "name": "Alice",
          "email": "alice@aol.com",
          "roles": [
            "user"
          ]
        }
      ]
    ]
  }
}"
`;

exports[`modifiers updateMapValue with a predicate fn should update a value of the map 1`] = `
"{
  "users": {
    "dataType": "Map",
    "value": [
      [
        1,
        {
          "name": "New Bob",
          "email": "bob@aol.com",
          "roles": [
            "user"
          ]
        }
      ],
      [
        2,
        {
          "name": "Alice",
          "email": "alice@aol.com",
          "roles": [
            "user"
          ]
        }
      ]
    ]
  }
}"
`;
